src/us/kbase/workspace/database/provenance/CreditMetadata.java

```diff
@@ -0,0 +1,345 @@
+				this.resourceType = DEFAULT_RESOURCE_TYPE;
+			}
+
+			try {
+				this.identifier = Common.checkPid(identifier, "identifier", false);
+			} catch (IllegalArgumentException e) {
+				this.errorList.add(e.getMessage());
+			}
+
+			// must be at least one contributor
+			this.contributors = Common.dedupeSimpleList(contributors);
+			if (this.contributors == null || this.contributors.isEmpty()) {
+				this.errorList.add("at least one contributor must be provided");
+			}
+
+			// must be at least one title
+			this.titles = Common.dedupeSimpleList(titles);
+			if (this.titles == null || this.titles.isEmpty()) {
+				this.errorList.add("at least one title must be provided");
+			}
+		}
+
+		/**
+		 * Sets comments for the resource
+		 * @param comments comments as a string
+		 * @return this builder
+		 */
+		public Builder withComments(final String comments) {
+			this.comments = Common.processMultilineString(comments);
+			return this;
+		}
+
+		/**
+		 * Sets the license for the resource
+		 * @param license license as a string
+		 * @return this builder
+		 */
+		public Builder withLicense(final String license) {
+			this.license = Common.processString(license);
+			return this;
+		}
+
+		/**
+		 * Sets the version
+		 * @param withVersion version string
+		 * @return this builder
+		 */
+		public Builder withVersion(final String version) {
+			this.version = Common.processString(version);
+			return this;
+		}
+
+		/**
+		 * Sets the dates of various resource life cycle events
+		 * @param withDates list of {@link EventDate}s.
+		 * @return this builder
+		 */
+		public Builder withDates(final List<EventDate> dates) {
+			this.dates = Common.dedupeSimpleList(dates);
+			return this;
+		}
+
+		/**
+		 * Sets the funding source(s)
+		 * @param withFunding list of {@link FundingReference}s.
+		 * @return this builder
+		 */
+		public Builder withFunding(final List<FundingReference> funding) {
+			this.funding = Common.dedupeSimpleList(funding);
+			return this;
+		}
+
+		/**
+		 * Sets the related identifiers
+		 * @param withRelatedIdentifiers list of {@link PermanentID}s.
+		 * @return this builder
+		 */
+		public Builder withRelatedIdentifiers(final List<PermanentID> relatedIdentifiers) {
+			this.relatedIdentifiers = Common.dedupeSimpleList(relatedIdentifiers);
+			return this;
+		}
+
+		/**
+		 * Build the {@link CreditMetadata}.
+		 *
+		 * @return the credit metadata.
+		 */
+		public CreditMetadata build() {
+
+			if (version == null && (dates == null || dates.isEmpty())) {
+				errorList.add("must provide either 'version' or one or more 'dates', ideally indicating when the resource was published or when it was last updated");
+			}
+
+			// If the license looks like a URL, ensure it's properly formatted and seems valid.
+			// For these purposes, if it starts with http or contains '://', assume it's an URL.
+			if (license != null && (license.toLowerCase().startsWith("http") || license.contains("://"))) {
```

Comment:
Why not just run processUrl if the string isn't null and check the protocol is http/s, and treat it as text otherwise?

Reply:
how is that advantageous over what is written here?

Reply:
Slightly simpler code - no need for the "does this look vaguely like a URL" check, just have processURL do it for you.

Also then you can do the equivalent of
```
if (!new HashSet<>(Arrays.asList("http", "https")).contains(url.getProtocol())) {
    // handle error
}
```

Reply:
Are you proposing something like this?

```java

if (license != null) {
	try {
		final URL licenseURL = Common.processURL(license, "license");
		// license field is a string, so convert the URL back to string form
		license = licenseURL.toURI().normalize().toString();
		if (!new HashSet<>(Arrays.asList("http", "https")).contains(licenseURL.getProtocol())) {
			// handle error
		}
	} catch (Exception e) {
		// treat as a string
		license = license.trim();
	}

```
If I'm understanding your suggestion right, that method doesn't differentiate between an invalid URL and a text string, though.

Reply:
Ok, I see, you want the `looks vaguely like a url check` to cause the method to reject things that aren't urls but look like urls.

It seems unlikely that license text starts with HTTP[S], but it seems pretty likely that there might be another url within the text, so I'm not sure how helpful the `://` check is. Maybe just check it starts with `http://` or `https://`?

Reply:
New set up: if it matches `^[a-zA-Z]+:+/+`, treat it like an URL. Otherwise, leave as-is. Covers typos in the protocol and colon-slash part (or valid alternate protocols like ftp), and the intelligent user will be able to deduce that multiple URLs can be entered by preceding them with some textual filler that doesn't match the afore-mentioned regex.

Reply:
I'm a bit iffy about this but not enough to actually do anything other than comment I'm iffy, so

:+1: